Update wiki (#992)
SonarSource/sonar-scanner-msbuild: docs/Wiki/ProductAndTestProjects.md

> ﻿This page refers to the SonarScanner for .NET (S4NET), also known as SonarScanner for MSBuild (S4MSB). You can read more about how to install and use it in the ["SonarScanner for .NET documentation"](https://redirect.sonarsource.com/doc/install-configure-scanner-msbuild.html).

Reviewer: S4NET seems a bit weird, we're missing the DOT (.) somehow.

Author: It's not perfect, but it's the best of available options I can see.
`S4MSB` - doesn't reflect current name anymore
`S4.NET` - I don't see the Scanner part here, looks more like URL or some .NET library
`S4DOTNET` - looses the advantage of a short abbreviation
`S4DNET` - doesn't remind the dot

Is there some better option?

Reviewer: My first thought was maybe S4.NET or S4DN but yep, this sounds a bit weird. We can even call it "SANETTT" with that, funny !

* any third-party analyzers referenced by the MSBuild project will be executed as part of the build (e.g. via NuGet package references). Issues from those analyzers will be uploaded to SonarQube/SonarCloud as [External Issues](https://sonarcloud.io/documentation/analysis/external-issues/). You can configure the third-party analyzer rules as normal using a `Ruleset`. During the build, S4NET will merge your custom ruleset with a ruleset generated from the Quality Profile. In the event of a conflict, the settings in the generated ruleset take precedence.

### Differences between analysis of test projects and product projects

The S4NET analyses MSBuild projects containing product code differently from projects containing test code.

#### Analysis of product projects

* analysis rules will be run against product projects, and the issues raised will be uploaded to SonarQube/SonarCloud.

> * only rules related to product code will execute.

Reviewer: i would prefer "will be executed" not sure how more correct it is.

* metrics and Lines of Code ("LOC") limit for commercial versions of SonarQube or private SonarCloud projects are calculated.

#### Analysis of test projects

* analysis rules will be run against test projects, unless excluded (see below), and the issues raised will be uploaded to SonarQube/SonarCloud.

> * only rules related to test code will execute.

Reviewer: i would prefer "will be executed" not sure how more correct it is.

> * test projects do not count towards to the Lines of Code ("LOC") limit for commercial versions of SonarQube or private SonarCloud projects.

Reviewer: Is it still correct ?

Author: Yes, test projects are still not included in LOC to avoid sudden licensing surprises on client side

Reviewer: Ok, is that a Scanner engine behavior ? I wasn't aware of that

Author: Yes, it's on Scanner side. See this:
https://github.com/SonarSource/sonar-scanner-msbuild/blob/0779f4d58eec352021ce36e32fe8df574c78bba8/src/SonarScanner.MSBuild.Tasks/Targets/SonarQube.Integration.targets#L353-L365

Until recently, analyzer didn't know what Scanner considered as a test code. So Analyzer created the `metrics.pb` file containing LOC and other stats for every project. And this target file then overwrites the output binary file to zero-length file (hacky).

In the next release of analyzer, we already can detect a test project from Scanner side and we'll not generate the file to save performance.

Author: Line 359 and 364 do the trick in the link above

﻿This page refers to the SonarScanner for .NET (S4NET), also known as SonarScanner for MSBuild (S4MSB). You can read more about how to install and use it in the ["SonarScanner for .NET documentation"](https://redirect.sonarsource.com/doc/install-configure-scanner-msbuild.html).

### Analyzer references

* the `SonarAnalyzer.CSharp`/`SonarAnalyzer.VB` analyzers will be added to the build even if those analyzers are not referenced in the MSBuild project (S4NET will add references to those analyzers on the fly during the build). The set of rules that is executed is determined by the project type and Quality Profile for the Sonar project in SonarQube/SonarCloud.

> * any third-party analyzers referenced by the MSBuild project will be executed as part of the build (e.g. via NuGet package references). Issues from those analyzers will be uploaded to SonarQube/SonarCloud as [External Issues](https://sonarcloud.io/documentation/analysis/external-issues/). You can configure the third-party analyzer rules as normal using a `Ruleset`. During the build, S4NET will merge your custom ruleset with a ruleset generated from the Quality Profile. In the event of a conflict, the settings in the generated ruleset take precedence.

Reviewer 2: I think we should document the fact that `.editorconfig` can be used, as well, for 3rd party analyzers. Not sure on its behavior with rulesets , though.

Author: As it's a Wiki, update is not bound to a release. We can improve it anytime. In this case, after Čaba finishes his investigation on the topic.

### Differences between analysis of test projects and product projects

The S4NET analyses MSBuild projects containing product code differently from projects containing test code.

#### Analysis of product projects

* analysis rules will be run against product projects, and the issues raised will be uploaded to SonarQube/SonarCloud.
* only rules related to product code will execute.

> * metrics and Lines of Code ("LOC") limit for commercial versions of SonarQube or private SonarCloud projects are calculated.

Reviewer 2: and also copy-paste detection

#### Analysis of test projects

* analysis rules will be run against test projects, unless excluded (see below), and the issues raised will be uploaded to SonarQube/SonarCloud.
* only rules related to test code will execute.
* test projects do not count towards to the Lines of Code ("LOC") limit for commercial versions of SonarQube or private SonarCloud projects.
* metrics are not calculated for test projects, although syntax colourisation and symbol highlighting are supported.

#### Analysis of excluded test projects

> * test projects are excluded from analysis on SonarQube up to version 8.8

Reviewer 2: ```suggestion
* test projects are excluded from analysis on SonarQube up to version 8.8.
```

* test projects can be excluded from analysis on SonarCloud and SonarQube from version 8.9 by adding `/d:sonar.dotnet.excludeTestProjects=true` S4NET parameter.

> * analysis rules are not run against excluded test projects i.e. no issues will be reported to SonarQube/SonarCloud. This is the case even if the test project references third-party NuGet analyzer packages - those analyzers will not be executed.

Reviewer 2: ```suggestion
* if test projects are excluded, no analyzers will be executed on any test project, i.e. no issues will be reported to SonarQube/SonarCloud. This is the case even if the test project references third-party NuGet analyzer packages - those analyzers will not be executed.
```

Author: I find the proposed version confusing. We're in `Analysis of excluded test projects` section. 
The confusing part:
`if test projects are excluded` - If? Is there a way to not-exclude some when I exclude them? 

I'll put `all test projects` on the previous points to make it more explicit.

> * excluded test projects do not count towards to the Lines of Code ("LOC") limit for commercial versions of SonarQube or private SonarCloud projects.

Reviewer 2: this can be removed - it's already mentioned in the "included" test projects section - and it's obvious that if they are excluded, they won't count

> * metrics are not calculated for test projects, although syntax colourisation and symbol highlighting are supported.

Reviewer 2: this can be removed
